src/main/java/beans/menus/EntityListResponse.java

```diff
@@ -156,27 +161,59 @@ private boolean matchEntity(Entity entity, String searchText) {
         return false;
     }
 
-    private Entity[] generateEntities(EntityScreen screen, Vector<TreeReference> references, EvaluationContext ec) {
-        Entity[] entities = new Entity[references.size()];
+    private EntityBean[] generateEntities(EntityScreen screen, Vector<TreeReference> references, EvaluationContext ec) {
+
+        List<Entity<TreeReference>> entityList = buildEntityList(screen.getShortDetail(), ec, references);
+        sort(entityList, screen.getShortDetail());
+
+        EntityBean[] entities = new EntityBean[entityList.size()];
         int i = 0;
-        for (TreeReference entity : references) {
-            Entity newEntity = processEntity(entity, screen, ec);
-            EntityDetailSubscreen[] subscreens = processDetails(screen, ec, entity);
+        for (Entity<TreeReference> entity : entityList) {
+            TreeReference treeReference = entity.getElement();
+            EntityBean newEntityBean = processEntity(treeReference, screen, ec);
+            EntityDetailSubscreen[] subscreens = processDetails(screen, ec, treeReference);
             if (subscreens != null) {
                 EntityDetailResponse[] responses = new EntityDetailResponse[subscreens.length];
                 for (int j = 0; j < subscreens.length; j++) {
                     responses[j] = new EntityDetailResponse(subscreens[j]);
                     responses[j].setTitle(subscreens[j].getTitles()[j]);
                 }
-                newEntity.setDetails(responses);
+                newEntityBean.setDetails(responses);
             }
-            entities[i] = newEntity;
+            entities[i] = newEntityBean;
             i++;
         }
         return entities;
     }
 
-    private Entity processEntity(TreeReference entity, EntityScreen screen, EvaluationContext ec) {
+    private List<Entity<TreeReference>> buildEntityList(Detail shortDetail, EvaluationContext context, Vector<TreeReference> references) {
+        NodeEntityFactory nodeEntityFactory = new NodeEntityFactory(shortDetail, context);
+        List<org.commcare.cases.entity.Entity<TreeReference>> full = new ArrayList<>();
+        for (TreeReference reference: references) {
+            full.add(nodeEntityFactory.getEntity(reference));
+        }
+        return full;
+    }
+
+    private void sort (List<Entity<TreeReference>> entityList, Detail shortDetail) {
```

Comment:
i don't really get how this works. i would have expected some sort of sort that would take place in the object fetch, not after the objects have been retrieved from the database. possible i'm completely misinterpreting what this code is doing. if so might be good to add a comment!